### PR TITLE
Feat add nested sort filter

### PR DESCRIPTION
**Pull request by PascalSenn**
- Aligns sorting with filters 
- Adds directives on sort fields
- Adds nested sorting 
<More detailed descirption to come>

#Todos 
- [x] Review the changes 
- [ ]  ~Fix duplicated name bugs (like #1105)~ moved to issue #1194

### CODE REVIEW

**Comment by PascalSenn** [comment]
@michaelstaib  

Currently the following test is failing

`public void Create_Explicit_Sorting_DifferentDescirptorOfSameType()`

This test creates a SortType for the same base type. 
```
  d.BindFieldsExplicitly()
                      .SortableObject(f => f.Baz1)
                      .Type(x => x.BindFieldsExplicitly().Sortable(y => y.BarProperty)
                            );
                    d.SortableObject(f => f.Baz2)
                    .Type(x => x.BindFieldsExplicitly()
                    .Sortable(y => y.BazProperty)
                        );
```

After considering it quite a bit I think this test should fail. 

We probably should not create names like BazSortingType and BazSortingType2. I guess if this happens then the developer has done something wrong. He did not consider that these two types are different. 
If we do this we will break input types which then could break existing queries.

**Comment by PascalSenn** [comment]
@michaelstaib  before we merge this we need the conventions

**Comment by michaelstaib** [comment]
Is this one ready for review?

**Comment by PascalSenn** [comment]
@michaelstaib  Ready for review :)

File: src/Core/Types.Sorting/ISortObjectOperationDescriptor.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        /// </param>
+        ISortObjectOperationDescriptor Name(NameString value);
+
+        /// <summary>
+        /// Ignore the specified property.
+        /// </summary>
+        /// <param name="property">The property that shall be ignored.</param>
+
+        ISortObjectOperationDescriptor Ignore();
+
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/ISortObjectOperationDescriptor.cs]: Whitespaces....

File: src/Core/Types.Sorting/ISortObjectOperationDescriptor~1.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        /// </param>
+        ISortObjectOperationDescriptor<TObject> Name(NameString value);
+
+        /// <summary>
+        /// Ignore the specified property.
+        /// </summary>
+        /// <param name="property">The property that shall be ignored.</param>
+
+        ISortObjectOperationDescriptor<TObject> Ignore();
+
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/ISortObjectOperationDescriptor~1.cs]: Whitespaces...

File: src/Core/Types.Sorting/ISortObjectOperationDescriptor~1.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        ISortObjectOperationDescriptor<TObject> Type(
+            Action<ISortInputTypeDescriptor<TObject>> descriptor);
+
+
+        /// <summary>
+        /// Specifies the type of the filter operation
+        /// </summary>
+        ISortObjectOperationDescriptor<TObject> Type<TFilter>()
+            where TFilter : SortInputType<TObject>;
+
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/ISortObjectOperationDescriptor~1.cs]: Whitespaces ...

File: src/Core/Types.Sorting/ISortObjectOperationDescriptor~1.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+        /// <summary>
+        /// Specifies the type of the filter operation
+        /// </summary>
+        /// <param name="descriptor"></param>
+        /// <returns></returns>
+        ISortObjectOperationDescriptor<TObject> Type(
+            Action<ISortInputTypeDescriptor<TObject>> descriptor);
+
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/ISortObjectOperationDescriptor~1.cs]: Whitespaces ...

File: src/Core/Types.Sorting/ISortOperationDescriptor.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        ISortOperationDescriptor Name(NameString value);
+
+        /// <summary>
+        /// Ignore the specified property.
+        /// </summary>
+        /// <param name="property">The property that shall be ignored.</param>
+
+        ISortOperationDescriptor Ignore();
+
+
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/ISortOperationDescriptor.cs]: Whitespaces ...

File: src/Core/Types.Sorting/ISortOperationDescriptor.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
+using HotChocolate.Language;
+
+namespace HotChocolate.Types.Sorting
+{
+    public interface ISortOperationDescriptor
+    {
+
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/ISortOperationDescriptor.cs]: Whitespaces ...

File: src/Core/Types.Sorting/SortObjectOperationDescriptor.cs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+        public new static SortObjectOperationDescriptor New(
+            IDescriptorContext context,
+            NameString name,
+            ITypeReference type,
+            SortOperation operation) =>
+            new SortObjectOperationDescriptor(context, name, type, operation);
+
+        public new static SortObjectOperationDescriptor CreateOperation(
+   PropertyInfo property,
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/SortObjectOperationDescriptor.cs]: Indentation ...

File: src/Core/Types.Sorting/SortObjectOperationDescriptor.cs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+                typeof(SortInputType<>).MakeGenericType(type),
+                TypeContext.Input);
+
+            return SortObjectOperationDescriptor.New(
+                context,
+                name,
+                typeReference,
+                operation);
+        }
+
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/SortObjectOperationDescriptor.cs]: Whitespaces ...

File: src/Core/Types.Sorting/SortObjectOperationDescriptor.cs [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        : SortOperationDescriptor
+        , ISortObjectOperationDescriptor
+    {
+
+        protected SortObjectOperationDescriptor(
+            IDescriptorContext context,
+            NameString name,
+            ITypeReference type,
+            SortOperation operation
+            ) : base(context, name, type, operation)
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/SortObjectOperationDescriptor.cs]: Brace has to go up a line ....

operation)
: base(....

File: src/Core/Types.Sorting/SortObjectOperationDescriptor.cs [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+using HotChocolate.Language;
+using HotChocolate.Types.Descriptors;
+
+namespace HotChocolate.Types.Sorting
+{
+    public class SortObjectOperationDescriptor
+        : SortOperationDescriptor
+        , ISortObjectOperationDescriptor
+    {
+
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/SortObjectOperationDescriptor.cs]: Whitespaces ....

File: src/Core/Types.Sorting/SortObjectOperationDescriptor.cs [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        }
+
+        public new ISortObjectOperationDescriptor Directive(
+            NameString name,
+            params ArgumentNode[] arguments)
+        {
+            base.Directive(name, arguments);
+            return this;
+        }
+
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/SortObjectOperationDescriptor.cs]: Whitespaces ....

File: src/Core/Types.Sorting/SortObjectOperationDescriptor~1.cs [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+        public new ISortObjectOperationDescriptor<TObject> Directive(
+            NameString name,
+            params ArgumentNode[] arguments)
+        {
+            base.Directive(name, arguments);
+            return this;
+        }
+
+        public ISortObjectOperationDescriptor<TObject> Type(Action<ISortInputTypeDescriptor<TObject>> descriptor)
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/SortObjectOperationDescriptor~1.cs]: line limit is 100 characters.

File: src/Core/Types.Sorting/SortObjectOperationDescriptor~1.cs [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+            return this;
+        }
+
+        public new static SortObjectOperationDescriptor<TObject> New(
+            IDescriptorContext context,
+            NameString name,
+            ITypeReference type,
+            SortOperation operation) =>
+            new SortObjectOperationDescriptor<TObject>(context, name, type, operation);
+
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/SortObjectOperationDescriptor~1.cs]: Whitespaces ...

File: src/Core/Types.Sorting/SortOperation.cs [hunk 14]
```diff
@@ -6,17 +6,18 @@ namespace HotChocolate.Types.Sorting
...
         }
-
-        public IEnumerable<SortOperationKind> AllowedSorts { get; }
+        public SortOperation(PropertyInfo property, bool isObject)
+        {
+            IsObject = isObject;
+            Property = property
+                       ?? throw new ArgumentNullException(nameof(property));
+        }
+        public bool IsObject { get; }
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/SortOperation.cs]: Formatting ...

File: src/Core/Types.Sorting/SortOperation.cs [hunk 15]
```diff
@@ -6,17 +6,18 @@ namespace HotChocolate.Types.Sorting
...
     public class SortOperation
     {
-        public SortOperation(
-            IEnumerable<SortOperationKind> allowedSorts,
-            PropertyInfo property)
+        public SortOperation(PropertyInfo property) : this(property, false)
         {
-            AllowedSorts = allowedSorts
-                           ?? throw new ArgumentNullException(nameof(allowedSorts));
             Property = property
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/SortOperation.cs]: Indentation

File: src/Core/Types.Sorting/SortOperationDescriptor.cs [hunk 16]
```diff
@@ -1,17 +1,22 @@
...
         : ArgumentDescriptorBase<SortOperationDefintion>
+        , ISortOperationDescriptor
     {
         protected SortOperationDescriptor(
-            IDescriptorContext context,
-            NameString name,
-            ITypeReference type,
-            SortOperation operation)
-            : base(context)
+         IDescriptorContext context,
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/SortOperationDescriptor.cs]: Indentation

File: src/Core/Types.Sorting/SortOperationDescriptor.cs [hunk 17]
```diff
@@ -23,9 +28,44 @@ public class SortOperationDescriptor
...
+            Definition.Ignore = true;
+            return this;
+        }
+
+        public ISortOperationDescriptor Name(NameString value)
         {
             Definition.Name = value.EnsureNotEmpty(nameof(value));
+            return this;
+        }
+
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/SortOperationDescriptor.cs]: Whitespaces

File: src/Core/Types.Sorting/SortOperationDescriptor.cs [hunk 18]
```diff
@@ -34,7 +74,29 @@ protected void Name(NameString value)
...
             SortOperation operation) =>
             new SortOperationDescriptor(
-                context, name, type, operation);
+          context, name, type, operation);
+
+        public static SortOperationDescriptor CreateOperation(
+            PropertyInfo property,
+            IDescriptorContext context)
+        {
+
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/SortOperationDescriptor.cs]: Whitespaces

File: src/Core/Types.Sorting/SortQueryableClosure.cs [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+    public class SortQueryableClosure
+    {
+        public SortQueryableClosure(Type type, string parameterName)
+        {
+            Parameter = Expression.Parameter(type, parameterName);
+            Instance = new Stack<Expression>();
+
+            Instance.Push(Parameter);
+        }
+        public ParameterExpression Parameter;
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/SortQueryableClosure.cs]: Reformat this document (line breaks)

File: src/Core/Types.Sorting/SortVisitorBase.cs [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+                if (inputType.Fields.TryGetField(node.Name.Value,
+                    out IInputField field))
+                {
+                    Operations.Push(field);
+                    Types.Push(field.Type);
+                    return VisitorAction.Continue;
+                }
+
+                throw new InvalidOperationException(
+                                SortingResources.SortObjectTypeFieldVisitor_InvalidType
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/SortVisitorBase.cs]: Indentation

File: src/Core/Types.Sorting/SortVisitorBase.cs [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+                );
+            }
+            else
+            {
+                throw new InvalidOperationException(
+                                SortingResources.SortObjectTypeVisitor_InvalidType
```
Comment (michaelstaib) [inline src/Core/Types.Sorting/SortVisitorBase.cs]: Indentation

File: src/Core/Types.Sorting.Mongo.Tests/MongoSortingObjectTests.cs [hunk 22]
```diff
@@ -0,0 +1,244 @@
...
+
+            ISchema schema = SchemaBuilder.New()
+                .AddQueryType<QueryType>()
+                .AddServices(serviceCollection.BuildServiceProvider())
+                .Create();
+
+            IQueryExecutor executor = schema.MakeExecutable();
+
+            IReadOnlyQueryRequest request = QueryRequestBuilder.New()
+                .SetQuery("{ items { model {foo} } }")
```
Comment (michaelstaib) [inline src/Core/Types.Sorting.Mongo.Tests/MongoSortingObjectTests.cs]: reformat query

File: src/Core/Types.Sorting.Mongo.Tests/MongoSortingObjectTests.cs [hunk 23]
```diff
@@ -0,0 +1,244 @@
...
+
+            ISchema schema = SchemaBuilder.New()
+                .AddQueryType<QueryType>()
+                .AddServices(serviceCollection.BuildServiceProvider())
+                .Create();
+
+            IQueryExecutor executor = schema.MakeExecutable();
+
+            IReadOnlyQueryRequest request = QueryRequestBuilder.New()
+                .SetQuery("{ items(order_by: { model: {foo: DESC }}) { model {foo }} }")
```
Comment (michaelstaib) [inline src/Core/Types.Sorting.Mongo.Tests/MongoSortingObjectTests.cs]: reformat query

File: src/Core/Types.Sorting.Mongo.Tests/MongoSortingObjectTests.cs [hunk 24]
```diff
@@ -0,0 +1,244 @@
...
+
+            ISchema schema = SchemaBuilder.New()
+                .AddQueryType<QueryType>()
+                .AddServices(serviceCollection.BuildServiceProvider())
+                .Create();
+
+            IQueryExecutor executor = schema.MakeExecutable();
+
+            IReadOnlyQueryRequest request = QueryRequestBuilder.New()
+                .SetQuery("{ paging(order_by: { model: {foo: DESC } }) { nodes { model {foo } } } }")
```
Comment (michaelstaib) [inline src/Core/Types.Sorting.Mongo.Tests/MongoSortingObjectTests.cs]: reformat query / line length ok?

File: src/Core/Types.Sorting.Mongo.Tests/MongoSortingObjectTests.cs [hunk 25]
```diff
@@ -0,0 +1,244 @@
...
+
+            ISchema schema = SchemaBuilder.New()
+                .AddQueryType<QueryType>()
+                .AddServices(serviceCollection.BuildServiceProvider())
+                .Create();
+
+            IQueryExecutor executor = schema.MakeExecutable();
+
+            IReadOnlyQueryRequest request = QueryRequestBuilder.New()
+                .SetQuery("{ items(order_by: { model:{ qux: DESC }}) { model{bar }} }")
```
Comment (michaelstaib) [inline src/Core/Types.Sorting.Mongo.Tests/MongoSortingObjectTests.cs]: reformat query

**Comment by michaelstaib** [comment]
@PascalSenn so can I merge it when I have fixed the formatting?

**Comment by michaelstaib** [comment]
It seems the integration tests are breaking.

**Pull request merged**